src/compiler/scala/tools/nsc/backend/jvm/opt/InlineInfoAttribute.scala

```diff
@@ -0,0 +1,139 @@
+ * However, an attribute allows us to save many bits. In particular, note that the strings in an
+ * InlineInfo are serialized as references to constants in the constant pool, and those strings
+ * (traitImplClassSelfType, method names, method signatures) would exist in there anyway. So the
+ * ScalaInlineAttribute remains relatively compact.
+ */
+case class InlineInfoAttribute(inlineInfo: InlineInfo) extends Attribute(InlineInfoAttribute.attributeName) {
+  /**
+   * Not sure what this method is good for, it is not invoked anywhere in the ASM framework. However,
+   * the example in the ASM manual also overrides it to `true` for custom attributes, so it might be
+   * a good idea.
+   **/
+  override def isUnknown: Boolean = false
+
+  /**
+   * Serialize the `inlineInfo` into a byte array. Strings are added to the constant pool and serialized
+   * as references.
+   */
+  override def write(cw: ClassWriter, code: Array[Byte], len: Int, maxStack: Int, maxLocals: Int): ByteVector = {
+    val result = new ByteVector()
+
+    result.putByte(InlineInfoAttribute.VERSION)
+
+    var hasSelfIsFinal = 0
+    if (inlineInfo.isEffectivelyFinal)               hasSelfIsFinal |= 1
+    if (inlineInfo.traitImplClassSelfType.isDefined) hasSelfIsFinal |= 2
+    result.putByte(hasSelfIsFinal)
+
+    for (selfInternalName <- inlineInfo.traitImplClassSelfType) {
+      result.putShort(cw.newUTF8(selfInternalName))
+    }
+
+    // The method count fits in a short (the methods_count in a classfile is also a short)
+    result.putShort(inlineInfo.methodInfos.size)
+
+    for ((nameAndType, info) <- inlineInfo.methodInfos) {
+      val (name, desc) = nameAndType.span(_ != '(')
+      // Name and desc are added separately because a NameAndType entry also stores them separately.
+      // This makes sure that we use the existing constant pool entries for the method.
+      result.putShort(cw.newUTF8(name))
+      result.putShort(cw.newUTF8(desc))
+
+      var inlineInfo = 0
+      if (info.effectivelyFinal)       inlineInfo |= 1
+      if (info.traitMethodNonAccessor) inlineInfo |= 2
+      if (info.annotatedInline)        inlineInfo |= 4
+      if (info.annotatedNoInline)      inlineInfo |= 8
+      result.putByte(inlineInfo)
+    }
+
+    result
+  }
+
+  /**
+   * De-serialize the attribute into an InlineInfo. The attribute starts at cr.b(off), but we don't
+   * need to access that array directly, we can use the `read` methods provided by the ClassReader.
+   *
+   * `buf` is a pre-allocated character array that is guaranteed to be long enough to hold any
+   * string of the constant pool. So we can use it to invoke `cr.readUTF8`.
+   */
+  override def read(cr: ClassReader, off: Int, len: Int, buf: Array[Char], codeOff: Int, labels: Array[Label]): InlineInfoAttribute = {
+    var next = off
+
+    val version = cr.readByte(next)            ; next += 1
+    assert(version == 1, s"unexpected version for the ScalaInlineInfo attribute: $version")
```

Comment:
What about if someone with the 2.11.6 optimizer depends on classfiles generated by 2.11.7 (assuming that version of the compiler emits version 2?). We should probably just `debugLog` that we can't understand the metadata rather than tripping an assertion.